### PR TITLE
Add bidirectional sync to admin and ban list

### PR DESCRIPTION
**Pull request by Cooldude2606**
TL;DR Adds factorio server to controller datastore and all other instances syncing of `user.isAdmin` and `user.isBanned`.

- When save patching is enabled the ban sync will use the ingame events to trigger an IPC which is then sent to all other instances. When it is disabled it will fallback to using log events generated by factorio.
- When save patching is enabled the admin sync will use the ingame events for "on map" players and the log events for "not on map" players. When it is disabled it will fallback to using log events for both on map and not on map players.
- Enough changes were made to allow `user.isWhitelisted` to also be synced, however the event is not currently triggered and is deemed out of scope for this PR because of the need to watch the whitelist json file and compare to previous versions.
- If a user does not exist (has never joined any server) and a sync event is triggered, then the user will be created.
- A dropdown list has been used for the config value to allow for sync events from an instance to be ignored. This also makes bidirectional sync an opt-in feature for those migrating from alpha 18. Bidirectional is the default for new instances.
- The `onPlayerEvent` plugin hook will be raised by the user sync events because of how similar it is in behaviour to join and leave events.

Out of Scope:
- Implement bidirectional whitelist sync using file watching.

~~A decision is needed on if these sync events will raise the `onPlayerEvent` plugin hook. A reason for doing so it because they are all log events relating directly to a single player which a plugin may wish to react to. By providing it through this hook it would abstract away the various methods of captuing these events manually. A reason for not doing this is because plugins can listen for updates on the user manager, although this would not indicate what has changed and so I am in favour of using the hook. An alternative is to create a new plugin hook for these events.~~
Edit: I have decided to add the logic to have 'onPlayerEvent' be raised by these new events.

## Changelog
```
### Features
- Add bidirectional sync to admin and ban list [#709](https://github.com/clusterio/clusterio/pull/709)

### Changes
- `onPlayerEvent` will now be raised by ban/unban/promote/demote/whitelisted/unwhitelisted [#709](https://github.com/clusterio/clusterio/pull/709)
```

### CODE REVIEW

**Comment by Danielv123** [comment]
Is this ready or is there anything we are waiting for?

**Comment by Cooldude2606** [comment]
The logic is ready, I plan to add tests tomorrow.

**Comment by Cooldude2606** [comment]
Patch coverage was missed because of the inclusion of dead code in preperation for the implementation of the out of scope whitelist sync. The test is present but is skipped due to the lack of implementation.

**Review by Danielv123 — APPROVED** [review]

File: packages/lib/src/config/definitions.ts [hunk 1]
```diff
@@ -481,19 +495,22 @@ export class InstanceConfig extends classes.Config<InstanceConfigFields> {
...
+			initialValue: "Bidirectional",
 		},
 		"factorio.sync_whitelist": {
-			description: "Synchronize whitelist with controller",
-			type: "boolean",
-			initialValue: true,
+			description: "Synchronize whitelist with the controller",
+			type: "string",
+			enum: ["Disabled", "Enabled"], // TODO: Implement bidirectional
+			initialValue: "Bidirectional",
```
[the remark this answers is not in the document]
Comment (Cooldude2606) [reply packages/lib/src/config/definitions.ts:507]: Ah I missed this when I made whitelist sync out of scope. Will update.

File: packages/host/src/Instance.ts [hunk 2]
```diff
@@ -787,16 +817,17 @@ rcon.print(game.table_to_json(players))`.replace(/\r?\n/g, " ");
 	async prepare() {
 		this.logger.verbose("Writing server-settings.json");
 		await this.writeServerSettings(true);
+		this._expectedUserUpdates = [];
 
-		if (this.config.get("factorio.sync_adminlist")) {
+		if (this.config.get("factorio.sync_adminlist") !== "Disabled") {
```
[the remark this answers is not in the document]
Comment (Cooldude2606) [reply packages/host/src/Instance.ts:822]: Rather than use a consttant dict, I have updated the type on the interface to be a union of strings. This will cause typescript to error if you attempt to test for a string not in the union.
